core/matrix/dense.cpp

```diff
@@ -1287,6 +1288,42 @@ void Dense<ValueType>::row_gather_impl(const Dense<ValueType>* alpha,
 }
 
 
+template <typename IndexType>
+size_type get_size(const array<IndexType>* arr)
+{
+    return arr->get_size();
+}
+
+template <typename IndexType>
+size_type get_size(const index_set<IndexType>* is)
+{
+    return is->get_num_elems();
+}
+
+
+template <typename ValueType, typename OutputType, typename IndexContainer>
+void row_scatter_impl(const IndexContainer* row_idxs,
+                      const Dense<ValueType>* orig, Dense<OutputType>* target)
+{
+    auto exec = orig->get_executor();
+    dim<2> expected_dim{static_cast<size_type>(get_size(row_idxs)),
+                        orig->get_size()[1]};
+    GKO_ASSERT_EQUAL_DIMENSIONS(expected_dim, orig);
+    GKO_ASSERT_EQUAL_COLS(orig, target);
+
+    bool invalid_access = false;
+
+    exec->run(dense::make_row_scatter(
+        make_temporary_clone(exec, row_idxs).get(), orig,
+        make_temporary_clone(exec, target).get(), invalid_access));
+
+    if (invalid_access) {
+        GKO_INVALID_STATE(
+            "Out-of-bounds access detected during kernel execution.");
```

Comment:
I think we can be more specific here
```suggestion
            "Out-of-bounds scatter index.");
```